### ISSUE TITLE
can we input x264 raw video to  mEncoder.onGetRgbaFrame(data, width, height) method

### ISSUE THREAD

**Issue by flyanra** [begeekmyfriend/yasea#291]
I need to stream a x264 video to rtmp server in flv format. I am trying to input the raw video data in the mEncoder.onGetRgbaFrame(data, width, height) method. Will it work. if not, is it possible to do. please suggest how can I do that. 

With some modifications, I removed the android camera input, and trying to input my raw x264 video to onGetRgbaFrame method.   When I run the code, the android device hangs.  How can i proceed ahead.

byte[] videoBuffer;   // array of x264 video
int size;     // size of the videoBuffer
int width;  // width of the frame
int height;  // height of the frame.

I am having the value of the above items. 

// In my Activity class I am doing
          mPublisher =  new SrsPublisher();
            mPublisher.setEnCoder(encoder);  
            mPublisher.setRtmpHandler(new RtmpHandler(this));
            mPublisher.setEncodeHandler(new SrsEncodeHandler(this));
            mPublisher.setRecordHandler(new SrsRecordHandler(this));
            mPublisher.setPreviewResolution(640, 480);
            mPublisher.setOutputResolution(640, 480);
            mPublisher.setVideoHDMode();
            mPublisher = new SrsPublisher(videoBuffer, width,height);;
            mPublisher.startPublish(rtmpUrl);

// and in  SrsPublisher class

   public SrsPublisher(byte[] data, int width, int height) {
        try{
            calcSamplingFps();
            mEncoder.onGetRgbaFrame(data, width, height);
        }catch (Exception e){
            e.printStackTrace();
        }
    }

**Comment by begeekmyfriend** [comment]
Yes you may. But the entry method is `onEncodedAnnexbFrame` and you have to make up the `MediaCodec.BufferInfo` object as a paremeter by yourself.

**Comment by flyanra** [comment]
**Hi begeekmyfriend,   I tried to use onEncodedAnnexbFrame  as suggested by you, here is the setup** 

byte[] videoBuffer; 	// array of x264 video
int size; 		// size of the videoBuffer
int width; 		// width of the frame (1280)
int height; 		// height of the frame.(720)

**I get the above parameters from a DJI drone camera.
I prepare the encoder like below**

        mSrsEncoder = new SrsEncoder(new SrsEncodeHandler(this));
        mSrsEncoder.setFlvMuxer(new SrsFlvMuxer(new RtmpHandler(this)));
        mSrsEncoder.setMp4Muxer(new SrsMp4Muxer(new SrsRecordHandler(this)));
        mSrsEncoder.setVideoSmoothMode();
        mSrsEncoder.setLandscapeResolution(ScreenWidth, ScreenHeight);
        mSrsEncoder.isSoftEncoder();
  
private void doStreamingWork(byte[] videoBuffer, int width, int height, long timeStamp) {
	if(videoBuffer!=null) {
                mPublisher = new SrsPublisher();
                mPublisher.setEnCoder(mSrsEncoder);
                mPublisher.setRtmpHandler(new RtmpHandler(this));
                mPublisher.setEncodeHandler(new SrsEncodeHandler(this));
                mPublisher.setRecordHandler(new SrsRecordHandler(this));
                mPublisher.setPreviewResolution(width, height);
                mPublisher.setOutputResolution(width, height);
                mPublisher.setVideoSmoothMode();
                mPublisher.setSrsPublisher(mSrsEncoder, videoBuffer, timeStamp, true);
                mPublisher.startPublish(rtmpUrl);
            }
 }




**I call the doStreamingWork   method when i get the video from drone Camera.**
mReceivedVideoDataCallBack = new DJICamera.CameraReceivedVideoDataCallback() {
            @Override
            public void onResult(final byte[] videoBuffer, int size) 

       			  doStreamingWork(mVideoBuffer, mCodecManager.getVideoWidth(), mCodecManager.getVideoHeight(), (oldSystemTime - System.nanoTime())*1000);
            }
        };                


**inside SrsPublisher Class**

public void setSrsPublisher(SrsEncoder encoder, byte[] data, long pts, boolean isKeyFrame) {
            mEncoder =  encoder;
            if(mEncoder!=null){
                calcSamplingFps();
                MediaCodec.BufferInfo vebi = new MediaCodec.BufferInfo();
                vebi.offset = 0;
                vebi.size = data.length;
                vebi.presentationTimeUs = pts;
                vebi.flags = true ? MediaCodec.BUFFER_FLAG_KEY_FRAME : 0;
                mEncoder.onEncodedAnnexbFrame(ByteBuffer.wrap(data), vebi);
                mEncoder.start();
            }
    }

**Inside SrsEncoder class**

 // when got encoded h264 es stream.
    public void onEncodedAnnexbFrame(ByteBuffer es, MediaCodec.BufferInfo bi) {
        try{        ByteBuffer record = es.duplicate();
            mp4Muxer.writeSampleData(videoMp4Track, record, bi);
            flvMuxer.writeSampleData(videoFlvTrack, es, bi);
        }catch (Exception e) {e.printStackTrace();}

    }

**Comment by flyanra** [comment]
I see that the Code is working. When i see at server, there is no Video but it shows that it is connected to mobile device. I do not know what I am doing wrong. The app shows toasts saying, connecting, connected to RTMP server. at the end link is established but there is  no video frames getting to server.

How can I fix this

**Comment by begeekmyfriend** [comment]
The connection packets are built in RTMP module itself and that does nothing with your code.

You said that the input as H.264 raw data, why did you still configure the encoder?

**Comment by flyanra** [comment]
Hi, 
Yes I have the input as H.264 raw data. I had to create a new instance of Encoder because It was null when I initialized the SrsPublisher object.  

Earlier Srs publisher object was set up to take input from Phone's camera using SrsCameraView. I replaced the SrsCameraView with my video Input view. After that I checked that encoder object is null.  So i had to create the encoder and initialize it.

**Comment by begeekmyfriend** [comment]
Then you need to check some sync bytes in `SrsFlvMuxer.java` such as annexB (0x0 0x0 0x0 0x1) or nalu header. You can use logcat for location.

**Comment by flyanra** [comment]
Hi Leo Ma, I checked for annexB errors and found that annexb is not matching. 
E/SrsFlvMuxer: annexb not match.
I/SrsFlvMuxer: 000-015: 0xbe 0x2 0x81 0x4e 0xfe 0xef 0xc5 0xab 0xbe 0x3b 0xa 0xd9 0xb1 0x75 0x1c 0x5b 

and the Toast shows 
Annexb not match for 4096B, pos=0;
The nalu_header is not null and its size is 4;

What does this error mean? What am I doing wrong? Thanks for your help.

**Comment by begeekmyfriend** [comment]
The H.264 stream from encoder typically follows annexB bytes. And then it will be replaced with 4-byte length value on muxing into container. So your raw H.264 stream is already follows length value bytes.

**Comment by flyanra** [comment]
That means that video I have is in proper format but I am doing something wrong somewhere else.
Can you Explain :-

Do I need to call mPublisher.setSrsPublisher(mSrsEncoder, videoBuffer, timeStamp, true) everytime I get a new frame or I have to call it only once. 
(Currently I am calling it everytime I get a new Video frame)

Do I need to call the mPublisher.startPublish()  everytime I get a new frame or I must call it only once.
(Currently I am calling it Only once, when I get the first video frame).

**Comment by begeekmyfriend** [comment]
What I mean is you need to remove some redundant code in `SrsFlvMuxer.java` to locate nalu header for each video frame. 

`flvMuxer.writeSampleData` is for every video frame. `mPublisher.startPublish` is for once.

**Comment by flyanra** [comment]
Okay, I will try to find what to remove. Thanks for the help. i will message to again.

**Comment by Timebutt** [comment]
Hi @flyanra, it seems you are trying to achieve the same thing I am researching right now, which is to stream the videofeed of a DJI drone over RTMP to a back-end server, correct?

Any updates on how far you've come this far? I'm struggling and could use any example code you have so far. I'm new to streaming video and trying to catch up fast, your help would be greatly appreciated!

**Comment by gabrielmadruga** [comment]
Hi @flyanra and @Timebutt, I want to do the same thing: stream the videofeed of a DJI drone.
Where you able to find a solution?
